### PR TITLE
Full Postgres 11+ feature support, subpartitioning support, and other enhancements

### PR DESCRIPTION
**Pull request by webandtech**
Thank you for this fantastic gem! I have incorporated this gem into a new project and made a good number of enhancements I'd like to contribute back to the project. I have ensured full backward compatibility to avoid a major version bump.

I've added `CHANGES.md`, as I think it could be useful if you were up for keeping it updated. Here's the summary for this PR:

#### Full Postgres 11 feature support has been added with complete backward compatibility
* When starting a fresh project with Postgres 11 or higher, you can disable template tables and enable primary key constraints on partitioned tables:
    * `config.create_template_tables` now governs whether template tables are created by default (defaults true).
    * `config.create_with_primary_key` passed down primary key options to CREATE TABLE, including support for composite primary keys
* `create_hash_partition`, `create_hash_partition_of`, and `attach_hash_partition` methods provide support for Hash partitions
* `create_default_partition_of` and `attach_default_partition` allows adding of default partitions to range and list partitioned tables
#### Full support for Subpartitioning
* `create_x_partition_of` methods now support `partition_type` and `partition_key`, which may be supplied to create
  a partitioned child table.
    * Use `create_x_partition_of` with the child table name to add a partition to your subpartition
    * Template tables are supported in that nested subpartitions will inherit from the top-level ancestor's template table, if found
* The `partitions` command now accepts an `include_subpartitions:` option which defaults to false for backward compatibility
    * Use `config.include_subpartitions_in_partition_list = true` to override the default
#### `add_index_on_all_partitions`
* Use this new adapter method in migrations to add an index on all partitions and subpartitions automatically
* This method supports `algorithm: :concurrently` to perform uptime operations, so even when using Postgres 11+ it is needed to avoid table locks.
* If you have many partitions, use the optional `in_threads:` option to parallelize index creation via the `parallel` gem
#### Minor enhancements
* Added adapter methods `partitions_for_table_name`, `parent_for_table_name`, and `table_partitioned?` to assist automating partition management, especially where subpartitions are involved

### CODE REVIEW

**Comment by rkrage** [comment]
Holy cow, thanks for doing this! I'll need a bit of time to go through this, but can hopefully get to it by the end of the week

**Comment by webandtech** [comment]
> Holy cow, thanks for doing this! I'll need a bit of time to go through this, but can hopefully get to it by the end of the week

No worries! I know it's a rather large PR... 😬

Happy to make changes once you've had a chance to review and possibly address codeclimate's issues (though that may require a refactor of `PgParty::AdapterDecorator` which might be better to do as part of a major version bump).

**Comment by rkrage** [comment]
Yeah, I think this whole project might be due for a refactor. The initial pattern of delegating to these decorator classes seemed like a good idea at the time. The idea was to minimize polluting core Rails classes with unnecessary methods. At least for the model side of things, I'm thinking we should have users explicitly include a `PgParty` module which would inject the methods directly into the class. This would avoid a ton of object allocations. I would definitely like to hear your thoughts, as it appears you're now _very_ familiar with the code base 😄

**Comment by rkrage** [comment]
Sorry I'm just getting around to reviewing this. Just curious, do you have a specific use case for subpartitioning? I just worry that we're adding a ton of complexity to support a feature that no one actually needs. Definitely open to discussion on that.

The rest of the additions are fantastic! More comments incoming...

File: lib/pg_party/adapter_decorator.rb [hunk 1]
```diff
@@ -26,17 +33,42 @@ def create_list_partition_of(table_name, values:, **options)
...
+      if partition_type
+        validate_supported_partition_type!(partition_type)
+        raise ArgumentError, '`partition_key` is required when specifying a partition_type' unless partition_key
+      end
 
-      validate_primary_key(primary_key)
+      like_option = if !partition_type || create_with_pks
+                      'INCLUDING ALL'
+                    else
+                      'INCLUDING ALL EXCLUDING INDEXES'
```
Comment (rkrage) [inline lib/pg_party/adapter_decorator.rb:62]: If we are excluding indexes when creating template tables, do template tables have any uses at this point? Should the create with PKs and template table config options be mutually exclusive?

File: lib/pg_party/adapter_decorator.rb [hunk 2]
```diff
@@ -63,49 +109,138 @@ def detach_partition(parent_table_name, child_table_name)
...
+
+      while (parents_parent = parent_for_table_name(parent)) do
+        parent = parents_parent
+      end
 
-      validate_primary_key(primary_key)
+      parent
+    end
+
+    def add_index_on_all_partitions(table_name, column_name, in_threads: nil, **options)
```
Comment (rkrage) [inline lib/pg_party/adapter_decorator.rb:144]: This is really only necessary for PG 10, correct?

File: lib/pg_party/adapter_decorator.rb [hunk 3]
```diff
@@ -120,6 +255,81 @@ def attach_partition(parent_table_name, child_table_name, constraint_clause)
...
+        end
+      else
+        _created_index_names << updated_name # Track as created before execution of concurrent index command
+        add_index_from_options(table_name, name: updated_name, type: index_type, algorithm: algorithm, using: using,
+                               columns: index_columns, options: index_options)
+      end
+
+      attach_child_index(updated_name, _parent_index_name) if _parent_index_name
+
+      return true if index_valid?(updated_name)
```
Comment (rkrage) [inline lib/pg_party/adapter_decorator.rb:290]: I was under the impression that AR would raise an error when index creation fails (even using the concurrent option). What's the reasoning for explicitly validating the index?

File: lib/pg_party/adapter_decorator.rb [hunk 4]
```diff
@@ -157,27 +367,69 @@ def quote_collection(values)
...
+
+      raise NotImplementedError, 'Default partitions are only available in Postgres 11 or higher'
     end
 
     def supports_partitions?
       __getobj__.send(:postgresql_version) >= 100000
     end
+
+    def postgres_11_plus?
+      __getobj__.send(:postgresql_version) >= 110000
```
Comment (rkrage) [inline lib/pg_party/adapter_decorator.rb:432]: Kind of a nitpick, but maybe we should add a `postgres_version` method

File: lib/pg_party/model/shared_methods.rb [hunk 5]
```diff
@@ -8,7 +8,10 @@ module SharedMethods
       def reset_primary_key
         if self != base_class
           base_class.primary_key
-        elsif partition_name = partitions.first
+        elsif (partitions = partitions(include_subpartitions: true)) && partitions.any?
```
Comment (rkrage) [inline lib/pg_party/model/shared_methods.rb:11]: I was never super thrilled that I was assigning a variable within this `elsif` statement. Now that it's a bit more complex maybe we should consider refactoring this method

File: lib/pg_party/model/shared_methods.rb [hunk 6]
```diff
@@ -8,7 +8,10 @@ module SharedMethods
...
+          partition_name = partitions.detect { |p| !connection.table_partitioned?(p) }
+          raise 'No child partitions exist for this model' unless partition_name
```
Comment (rkrage) [inline lib/pg_party/model/shared_methods.rb:13]: Just a note for myself. I think it's time to introduce `PgParty` specific error classes

File: lib/pg_party/model_decorator.rb [hunk 7]
```diff
@@ -62,15 +62,11 @@ def list_partition_key_in(*values)
...
-          FROM pg_tables
-          INNER JOIN pg_inherits
-            ON pg_tables.tablename::regclass = pg_inherits.inhparent::regclass
-          WHERE pg_tables.tablename = #{connection.quote(table_name)}
-        SQL
+    alias_method :hash_partition_key_in, :list_partition_key_in
+
+    def partitions(include_subpartitions: PgParty.config.include_subpartitions_in_partition_list)
+      PgParty.cache.fetch_partitions(cache_key, include_subpartitions) do
+        connection.partitions_for_table_name(table_name, include_subpartitions: include_subpartitions)
```
Comment (rkrage) [inline lib/pg_party/model_decorator.rb:69]: 👍 for moving SQL out of this class

File: spec/dummy/db/schema.rb [hunk 8]
```diff
@@ -116,6 +116,28 @@
...
+    :bigint_int_list_date_range_subpartitions_a,
+    name: :bigint_int_list_date_range_subpartitions_a_1,
+    start_range: Time.now - 1.day,
+    end_range: Time.now + 10.days
+
+  create_list_partition_of \
+    :bigint_int_list_date_range_subpartitions,
+    name: :bigint_int_list_date_range_subpartitions_b,
+    values: [3, 4]
+
```
Comment (rkrage) [inline spec/dummy/db/schema.rb:140]: Should we add some integration tests for hash partitioning? Or maybe you have and I missed that? I guess part of the problem is that we don't want to run the migration for PG 10.

**Comment by webandtech** [comment]
> Sorry I'm just getting around to reviewing this. Just curious, do you have a specific use case for subpartitioning? I just worry that we're adding a ton of complexity to support a feature that no one actually needs. Definitely open to discussion on that.
> 
> The rest of the additions are fantastic! More comments incoming...

Yes we have use cases for subpartitioning. Specifically when using HASH and LIST partition types together or with RANGE as well. Think of a system where you have 100,000 customers and you want to use HASH to distribute the data by default on `account_id`, then LIST subpartitions to allow for pulling out larger customers into their own partitions. The structure for doing this is:
```
HASH partition by `account_id`
  LIST partition by `account_id`
    DEFAULT partition
    partition for account_ids 5,6,7 (break accounts 5,6,7 into their own partition), possibly relocate to a foreign tables even
```

For LIST, RANGE this becomes useful as well... Say you have the above structure and you have a particularly large table called `events` which you want to partition by RANGE, you can further partition within the structure.

Postgres 12 includes a lot of performance enhancements to make the query planner performant with 1000s of partitions.

File: spec/dummy/db/schema.rb (hunk 8 above)
Comment (webandtech) [reply spec/dummy/db/schema.rb:140]: Yes, I added a spec `spec/integration/migration_postgres_11_plus_spec.rb`. When CI runs the suite on Postgres 11 and 12 it will run those tests. Works perfectly with your circleci setup that tests the gem on multiple PG versions... it was super helpful.

There is also a spec  `spec/integration/migration_postgres_10_spec.rb` which only runs in Postgres 10 and tests the errors being raised if you try to perform an unsupported action.

File: lib/pg_party/adapter_decorator.rb (hunk 3 above)
Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb:290]: The way that Postgres 11+ indices work on partitioned tables let me to this solution to ensure the whole operation succeeds. The process is as follows:
1. You create an index ON ONLY the parent (partitioned) table - this index is immediately marked invalid by postgres
2. You create all indices on all child partitions, repeating 1 recursively for subpartitions.
  a. After creating each index you ATTACH it to the parent index
3. Once all partitions have indices which are attached to the parent index, Postgres detects this and marks the parent index as valid. If you miss any partitions, the parent index is left in an invalid state and not used by the query planner

So basically, the validity check needed to ensure the entire hierarchy of indices was successful created and attached up the chain. If anything was missed, the validity check will fail and the method with unwind all the indices created via DROP INDEX IF EXISTS so it doesn't leave any orphaned indices.

Because the whole operation is done outside of a transaction when using CONCURRENTLY and also the support for parallel threads, I felt like this was an important check to add especially.

File: lib/pg_party/adapter_decorator.rb (hunk 2 above)
Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb:144]: The #1 use case for this method is to provide an algorithm for creating properly referenced indices on partitioned tables using CONCURRENTLY (which postgres 11/12 does not allow to be used when you create an index on a partitioned (parent) tables. If you don't do this and you add an index to a partitioned table in Postgres 11+ it will lock all the partitions while it builds the index. So, with this method, you can have a table with 100s or even 1000s of partitions, including multiple levels of subpartitions, and you can add an index to all of them conveniently without any table locks and take advantage of postgres 11+'s index propagation so if you add a new partition later it will pick up the index automatically.

See my other comment further down for more info.

`in_threads:` takes an integer and with CONCURRENTLY will spread the index creation across multiple connections which is helpful if you have a whole lot of partitions, especially if some of them are rather large tables. Concurrent index creation is rather slow...

File: lib/pg_party/adapter_decorator.rb (hunk 1 above)
Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb:62]: I think they do, 1) for preserving backward compatibility for folks already using them, and 2) because if you have expressions in your partition key you cannot use postgres 11/12's primary key support on partitioned tables.

Template tables are more flexible and I figure people may already be using them in Postgres 11+ but may want to use the native features on new tables and mix the features to get the best of both worlds... could be revisited in a version 2 of the gem...

File: lib/pg_party/model/shared_methods.rb (hunk 5 above)
Comment (webandtech) [reply lib/pg_party/model/shared_methods.rb]: Refactored this method in https://github.com/rkrage/pg_party/pull/54/commits/b1783ffe1b70217480c818fd67701a191ff90264

File: lib/pg_party/adapter_decorator.rb (hunk 4 above)
Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb]: Refactored in https://github.com/rkrage/pg_party/pull/54/commits/b1783ffe1b70217480c818fd67701a191ff90264 where I added a `postgres_major_version` method

File: lib/pg_party/model/shared_methods.rb [hunk 9]
```diff
@@ -6,13 +6,15 @@ module PgParty
...
-          base_class.primary_key
-        elsif partition_name = partitions.first
-          in_partition(partition_name).get_primary_key(base_class.name)
-        else
-          get_primary_key(base_class.name)
-        end
+        return base_class.primary_key if self != base_class
+
+        partitions = partitions(include_subpartitions: true)
+        return get_primary_key(base_class.name) unless partitions.any?
```
Comment (rkrage) [inline lib/pg_party/model/shared_methods.rb:12]: I _think_ `any?` will iterate over the entire array. Maybe we want `partitions.empty?` here?

File: lib/pg_party/adapter_decorator.rb [hunk 10]
```diff
@@ -63,49 +109,138 @@ def detach_partition(parent_table_name, child_table_name)
...
-      template         = options.fetch(:template, true)
-      id               = options.fetch(:id, :bigserial)
-      primary_key      = options.fetch(:primary_key) { calculate_primary_key(table_name) }
+    def parent_for_table_name(table_name, traverse: false)
+      parent = select_values(%[
+          SELECT pg_inherits.inhparent::regclass::text
+          FROM pg_tables
+          INNER JOIN pg_inherits
+            ON pg_tables.tablename::regclass = pg_inherits.inhrelid::regclass
+          WHERE pg_tables.tablename = #{quote(table_name)}
```
Comment (rkrage) [inline lib/pg_party/adapter_decorator.rb:133]: Did we want to add a clause to check `current_schema()` here as well?

File: lib/pg_party/adapter_decorator.rb (hunk 2 above)
Comment (rkrage) [reply lib/pg_party/adapter_decorator.rb:144]: Do you think it would be overkill to validate that the number of threads isn't higher than the connection pool? I could go either way on that. Curious what you think

Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb:146]: Yes, it seems like a good idea. I added a check in https://github.com/webandtech/pg_party/commit/194486cafc08dba50b9e119513eebfa03105435e which requires in_threads to be lower than connection pool size (there should at least one additional connection for the parent thread)

File: lib/pg_party/model/shared_methods.rb (hunk 9 above)
Comment (webandtech) [reply lib/pg_party/model/shared_methods.rb]: `any?` should return immediately when the first element is found, but I swapped it for `empty?` in https://github.com/webandtech/pg_party/commit/194486cafc08dba50b9e119513eebfa03105435e because it is also more readable

File: lib/pg_party/adapter_decorator.rb (hunk 10 above)
Comment (webandtech) [reply lib/pg_party/adapter_decorator.rb]: Done in https://github.com/webandtech/pg_party/commit/194486cafc08dba50b9e119513eebfa03105435e

**Comment by webandtech** [comment]
> Yeah, I think this whole project might be due for a refactor. The initial pattern of delegating to these decorator classes seemed like a good idea at the time. The idea was to minimize polluting core Rails classes with unnecessary methods. At least for the model side of things, I'm thinking we should have users explicitly include a `PgParty` module which would inject the methods directly into the class. This would avoid a ton of object allocations. I would definitely like to hear your thoughts, as it appears you're now _very_ familiar with the code base 😄

I think the way you approached this has advantages - specifically, it avoids conflicts in method naming except for the methods specifically injected, which is nice from a maintainability and refactoring perspective. I would probably spike on 2 ways to address the excess allocations you mentioned:

1. Identify if we can instantiate a decorator class once per model class and memoize it.
2. If 1 doesn't solve, possibly passing the reference to `self` more explicitly rather than relying on `SimpleDelegator` which must be instantiated each time

I only thought about it for a few minutes and it doesn't seem to be a problem for my use cases at the moment, but something I could help look at in the future perhaps. I'm also a fan of "if it ain't broke, don't fix it" so would wait until I see some real pain before trying to a approach a substantial refactor.

File: pg_party.gemspec [hunk 11]
```diff
@@ -24,6 +24,7 @@ Gem::Specification.new do |spec|
 
   spec.add_runtime_dependency "activerecord", ">= 5.0", "< 6.1"
   spec.add_runtime_dependency "ruby2_keywords", "~> 0.0.2"
+  spec.add_runtime_dependency "parallel"
```
Comment (rkrage) [inline pg_party.gemspec:27]: Can we add some tighter constraints on this dependency? Like `~> 1.x` (unsure what minimum version is required)

Comment (webandtech) [reply pg_party.gemspec]: Set it to `~> 1.0` in https://github.com/webandtech/pg_party/commit/b798ce01da8005bd91df6a36e3265e6dd28cc7a7

**Comment by rkrage** [comment]
Apologies for the delay. Working through some access issues with my RubyGems account 😞
